x-pack/plugins/security_solution/public/resolver/models/schema.ts

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+type Validator<T> = (value: unknown) => value is T;
+type TypeOf<V extends Validator<unknown>> = V extends Validator<infer T> ? T : never;
+
+/**
+ * Validate that `value` matches at least one of `validators`.
+ * Use this to create a predicate for a union type.
+ * e.g.
+ * ```
+ * import * as schema from './schema';
+ * const isAscOrDesc: (value: unknown) => value is 'asc' | 'desc' = schema.oneOf([
+ *   schema.literal('asc' as const),
+ *   schema.literal('desc' as const),
+ * ]);
+ * ```
+ */
+export function oneOf<V extends Array<Validator<unknown>>>(validators: V) {
+  return function (
+    value: unknown
+  ): value is V extends Array<Validator<infer ElementType>> ? ElementType : never {
+    for (const validator of validators) {
+      if (validator(value)) {
+        return true;
+      }
+    }
+    return false;
+  };
+}
+
+/**
+ * Validate that `value` is an array and that each of its elements matches `elementValidator`.
+ * Use this to create a predicate for an array type.
+ * ```
+ * import * as schema from './schema';
+ * const isAscOrDesc: (value: unknown) => value is 'asc' | 'desc' = schema.oneOf([
+ *   schema.literal('asc' as const),
+ *   schema.literal('desc' as const),
+ * ]);
+ * ```
+ */
+export function array<V extends Validator<unknown>>(elementValidator: V) {
+  return function (
+    value: unknown
+  ): value is Array<V extends Validator<infer ElementType> ? ElementType : never> {
+    if (Array.isArray(value)) {
+      for (const element of value as unknown[]) {
+        const result = elementValidator(element);
+        if (!result) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return false;
+  };
+}
+
+/**
+ * The keys of `T` where `undefined` is assignable to the corresponding value.
+ * Used to figure out which keys could be made optional.
+ */
+type KeysWithOptionalValues<T extends { [key: string]: unknown }> = {
+  [K in keyof T]: undefined extends T[K] ? K : never;
+}[keyof T];
+
+/**
+ * `T` with required keys changed to optional if the corresponding value could be `undefined`.
+ * Converts a type like `{  key: number | undefined; requiredKey: string }` to a type like `{ key?: number | undefined; requiredKey: string }`
+ * This allows us to write object literals that omit a key if the value can accept `undefined`.
+ */
+type OptionalKeyWhenValueAcceptsUndefined<T extends { [key: string]: unknown }> = {
```

Comment:
nice 👍

Reply:
Yeah, I feel like that one should go in TS standard